lib/internal/assert/snapshot.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+const {
+  ArrayPrototypeMap,
+  Boolean,
+  SafeMap,
+  SafeSet,
+  StringPrototypeIndexOf,
+  StringPrototypeSplit,
+  StringPrototypeSubstring,
+  StringPrototypeReplace,
+  Symbol,
+} = primordials;
+
+const { emitExperimentalWarning } = require('internal/util');
+const { codes: { ERR_ASSERT_SNAPSHOT_NOT_SUPPORTED } } = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const { inspect } = require('internal/util/inspect');
+const { createReadStream, createWriteStream } = require('fs');
+const path = require('path');
+const assert = require('assert');
+
+const kUpdateSnapshot = Boolean(process.env.NODE_UPDATE_SNAPSHOT);
+const kInitialSnapshot = Symbol('kInitialSnapshot');
+const kDefaultDelimiter = '\n#*#*#*$#*#*#*#*#*#*$#*#*#\n';
+
+function defaultSnapshotPath() {
+  if (process.mainModule) {
+    const { dir, name } = path.parse(process.mainModule.filename);
+    return path.join(dir, `${name}.snapshot`);
+  }
+  if (!process.argv[1]) {
```

Comment:
Is this to check if we're in the REPL? wouldn't it be better to just check if there is no `process.mainModule?.filename`?

Reply:
esm does not have `process.mainModule` either, so `argv[1]` is used instead

Reply:
I haven't read the code, but reminder that `process.argv[1] == false` when doing stdin eval and `--eval`, not just in REPL.

Reply:
Yes, it is documented to not be supported at this stage. 
In a further stage we can support the source and target options that were in the original commit